src/ringct/rctSigs.cpp

```diff
@@ -1600,17 +1603,24 @@ namespace rct {
 
       ::do_serialize(ar, const_cast<rctSig&>(rv));
 
-      crypto::hash h;
-      cryptonote::get_blob_hash(ss.str(), h);
+      const std::string ser_sig = ss.str();
+      const crypto::hash tx_hash = cryptonote::get_transaction_hash(tx);
+
+      crypto::hash sig_and_tx_hash;
+      KECCAK_CTX kctx; // not using cn_fast_hash here because we need 2 updates
```

Comment:
You could just do
```
ser_sig.append(reinterpret_cast<const char*>(&tx_hash), sizeof(tx_hash));
```
and then call `get_blob_hash` as usual. Keeping all keccak code in one place is better.

Reply:
Yes, the only problem is that this might waste an extra allocation. `keccak_init`, `keccak_update`, and `keccak_finish` are used other places in the codebase besides here, `wallet2.cpp` for instance. I also added a unit test to confirm that `cn_fast_hash` uses keccak.

Reply:
Yes, this code if good enough as is. Also, we shouldn't really care that `cn_fast_hash` is keccak because keccak here is not exposed to the calling code and can basically use any cryptographically secure hash here.

Reply:
Yeah I just added it in case we migrate away from keccak at any point, a unit test will remind us that we can also upgrade this hashing code instead of us having to remember.

Reply:
I'd like to discourage non-default use of crypto library functions unless there is a good reason. In this case, just reserve capacity in the string before appending anything, then call to `cn_fast_hash()` like normal.